From bcf7963c5a19f614b7eca4287636a9e0bf4c484b Mon Sep 17 00:00:00 2001
From: Jan Dominiak <janx.dominiak@intel.com>
Date: Thu, 15 Oct 2020 10:37:32 +0200
Subject: [PATCH] PONRTSYS-7970:

Updated subblock handling for PCE rules in trap action:
* subblock now have a name
* only one subblock will be created for all PCE rules for trapping, with maximum size of 32
---
 drivers/net/datapath/pon_qos/pon_qos_tc_trap.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c b/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
index 793563595f6e..fdae7f570311 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
@@ -38,10 +38,14 @@ struct pce_rule_storage {
 	struct list_head list;
 };
 
+#define TC_TRAP_SUBBLK_SIZE 32
+
 static void pon_qos_trap_set_default(struct net_device *dev,
 				     GSW_PCE_rule_t *pce_rule,
 				     struct dp_pce_blk_info *pce_blk_info)
 {
+	struct pce_rule_storage *p;
+
 	/* Set default values for trap action */
 	pce_rule->action.ePortMapAction = GSW_PCE_ACTION_PORTMAP_CPU;
 	pce_rule->pattern.bEnable = 1;
@@ -50,6 +54,17 @@ static void pon_qos_trap_set_default(struct net_device *dev,
 	pce_rule->region = GSW_PCE_RULE_COMMMON;
 
 	pce_blk_info->region = pce_rule->region;
+	pce_blk_info->info.subblk_size = TC_TRAP_SUBBLK_SIZE;
+	strncpy(pce_blk_info->info.subblk_name, "TRAP",
+		sizeof(pce_blk_info->info.subblk_name));
+
+	/* All PCE rules for trap action should be put in the same subblock,
+	 * take subblk_id from the first rule on the list. If there are no
+	 * rules in storage, leave subblk_id equal to 0, to make DP create new
+	 * subblock for trap action rules. */
+	list_for_each_entry(p, &tc_trap_storage, list) {
+		pce_blk_info->info.subblk_id = p->subblk_id;
+	}
 
 	return;
 }
